lang/rust/avro/src/schema_compatibility.rs

```diff
@@ -133,38 +156,64 @@ impl Checker {
             {
                 for field in r_fields.iter() {
                     if let Some(pos) = w_lookup.get(&field.name) {
-                        if !self.full_match_schemas(&w_fields[*pos].schema, &field.schema) {
-                            return false;
+                        if self
+                            .full_match_schemas(&w_fields[*pos].schema, &field.schema)
+                            .is_err()
+                        {
+                            return Err(Error::CompatibilityError(format!("Field {} in reader schema does not match the type in the writer schema", field.name)));
                         }
                     } else if field.default.is_none() {
-                        return false;
+                        return Err(Error::CompatibilityError(format!(
+                            "Field {} in reader schema must have a default value",
+                            field.name
+                        )));
                     }
                 }
             }
         }
-        true
+        Ok(())
     }
 
-    fn match_union_schemas(&mut self, writers_schema: &Schema, readers_schema: &Schema) -> bool {
+    fn match_union_schemas(
+        &mut self,
+        writers_schema: &Schema,
+        readers_schema: &Schema,
+    ) -> AvroResult<()> {
         let w_type = SchemaKind::from(writers_schema);
         let r_type = SchemaKind::from(readers_schema);
 
         assert_eq!(r_type, SchemaKind::Union);
 
         if w_type == SchemaKind::Union {
```

Comment:
maybe you could do this instead, to reduce the nesting

```rs
if w_type != SchemaKind::Union {
  return Err(Error::CompatibilityError(String::from(
                "readers_schema should have been Schema::Union",
            )));
}```

Reply:
I solved in a different way but this helped to see that some code was never called and the check `assert_eq!(r_type, SchemaKind::Union);` is actually not needed.